core/src/main/scala/kafka/server/metadata/PartitionMetadataProcessor.scala

```diff
@@ -121,6 +131,70 @@ class PartitionMetadataProcessor(kafkaConfig: KafkaConfig,
 
   // visible for testing
   private[metadata] class MetadataMgr(numBrokersAdding : Int, numTopicsAdding: Int, numBrokersFencing: Int) {
+    // support coalescing multiple config changes together and writing them on-demand
+    val configResourceChanges: mutable.Map[ConfigResource, mutable.Map[String, String]] = mutable.Map.empty
+
+    // must call writeConfigs() to apply these changes
+    def addConfigResourceChange(configResource: ConfigResource, name: String, value: String): Unit = {
+      if (configResource.`type`() == ConfigResource.Type.BROKER) {
+          name.toInt // check for valid broker ID and throw NumberFormatException if not
+      }
+      val changesForResource = configResourceChanges.getOrElse(configResource, mutable.Map.empty)
+      val firstChangeForResource = changesForResource.isEmpty
+      changesForResource(name) = value
+      if (firstChangeForResource) {
+        configResourceChanges.put(configResource, changesForResource)
+      }
+    }
+
+    // write all the coalesced config changes and return the affected resources (if any)
+    def writeConfigChangesIfNecessary(): Seq[ConfigResource] = {
+      if (configResourceChanges.isEmpty) {
+        List.empty
+      } else {
+        var configMapToWrite: Map[ConfigResource, Map[String, String]] = configMap
+        configResourceChanges.foreach { case (configResource, changedConfigsForResource) =>
+          val oldConfigsForResource: Map[String, String] = configMapToWrite.getOrElse(configResource, Map.empty)
+          val newConfigsForResource: Map[String, String] = oldConfigsForResource ++ changedConfigsForResource
```

Comment:
How are configs deleted?

Reply:
Thanks for asking that question.  I assume a value equal to the empty string implies a delete, and I'll have to react to that by deleting the config instead of literally setting the value to the empty string.

Reply:
cc: @cmccabe to confirm that a deleted config appears as a name-value pair with an empty value.

Reply:
It should be a name-value pair with a null value.  null and empty are not equivalent here, unfortunately.